plutus-tutorial/doc/plutus-tx.adoc

```diff
@@ -0,0 +1,383 @@
+to just write in place.
+
+== Typeclasses
+
+So far we have used functions like `lessThanEqInteger` for comparing `Integer` s,
+which is much less convenient than `<` from the standard Haskell `Ord` typeclass.
+
+Fortunately, Plutus Tx does support typeclasses. Unfortunately, we need to
+redefine the standard typeclasses do so, since we require the class methods to
+be `INLINABLE`, and the implementations for types such as `Integer` use the
+Plutus Tx builtins.
+
+Redefined versions of many standard typeclasses are available in the Plutus Tx
+Prelude. As such you should be able to use typeclass functions in your Plutus Tx
+programs successfully.
+
+For example, here is a version of the `pastEnd` function using `<`. This will be
+compiled to exactly the same code as the previous definition.
+
+[source,haskell]
+----
+-- | Check whether a given time is past the end date.
+pastEnd' :: CompiledCode (EndDate -> Integer -> Bool)
+pastEnd' = $$(compile [|| \(end::EndDate) (current::Integer) -> case end of
+    Fixed n -> n < current
+    Never -> False
+   ||])
+----
+
+== The Plutus Tx Prelude
+
+The `Language.PlutusTx.Prelude` module is a drop-in replacement for the
+normal Haskell Prelude, but with some functions and typeclasses redefined to be easier
+for the Plutus Tx compiler to handle (i.e. `INLINABLE`).
+
+You should use the Plutus Tx Prelude whenever you are writing code that you expect to compile with
+the Plutus Tx compiler. All of the definitions in the Plutus Tx Prelude have
+working Haskell definitions, so you can use them in normal Haskell code too,
+although the Haskell Prelude versions will probably perform better.
+
+To use the Plutus Tx Prelude, use the `NoImplicitPrelude` language
+pragma, and import `Language.PlutusTx.Prelude`.
+
+Plutus Tx has some builtin types and functions available for working
+with primitive data (integers and bytestrings), as well as a few special
+functions. These builtins are also exported from the Plutus Tx Prelude.
+
+The `error` builtin deserves a special mention. `error` causes the
+transaction to abort when it is evaluated, which one way to trigger
+validation failure.
+
+== Lifting values
+
+So far we’ve seen how to define pieces of code _statically_ (when you
+_compile_ your main Haskell program), but you are likely to want to generate code
+_dynamically_ (when you _run_ your main Haskell program). For example, you
+might be writing the body of a transaction to initiate a crowdfunding
+smart contract, which would need to be parameterized by data
+determining the size of the goal, the campaign start and end times, etc.
+
+We can do this in the same way that we normally parameterize code in functional
+programming: we write the static code as a _function_, and we provide the
+argument later to configure it.
+
+In our case we have a complication, in that we want to make the argument and
+apply the function to it at runtime. Plutus Tx provides a mechanism to do this
```

Comment:
It should be clearer that this is the runtime of the client code, not when the script is evaluated on the blockchain